File: backend/app/api/user_position_api.py
from flask import Blueprint, request, jsonify
from app.services.user_position_service import (
    get_all_positions, get_position_by_id, create_position, update_position, delete_position
)
from app.models.stock_company import StockCompany

bp = Blueprint('user_position', __name__, url_prefix='/api/user_position')

@bp.route('/<int:user_id>', methods=['GET'])
def list_positions(user_id):
    filters = {'user_id': user_id}
    ts_code = request.args.get('ts_code')
    if ts_code:
        filters['ts_code'] = ts_code
    page = int(request.args.get('page', 1))
    page_size = int(request.args.get('page_size', 20))
    query = get_all_positions(filters, query_only=True)
    total = query.count()
    items = query.offset((page - 1) * page_size).limit(page_size).all()
    result = []
    for i in items:
        d = i.as_dict()
        company = StockCompany.query.filter_by(ts_code=i.ts_code).first()
        d['stock_info'] = company.as_dict() if company else None
        result.append(d)
    return jsonify({'data': result, 'total': total})

@bp.route('/<int:user_id>/<int:position_id>', methods=['GET'])
def get_position(user_id, position_id):
    position = get_position_by_id(position_id)
    if not position or position.user_id != user_id:
        return jsonify({'error': 'Not found'}), 404
    return jsonify(position.as_dict())

@bp.route('/<int:user_id>', methods=['POST'])
def add_position(user_id):
    data = request.json
    data['user_id'] = user_id
    position = create_position(data)
    return jsonify(position.as_dict()), 201

@bp.route('/<int:user_id>/<int:position_id>', methods=['PUT'])
def edit_position(user_id, position_id):
    position = get_position_by_id(position_id)
    if not position or position.user_id != user_id:
        return jsonify({'error': 'Not found'}), 404
    data = request.json
    position = update_position(position_id, data)
    if not position:
        return jsonify({'error': 'Not found'}), 404
    return jsonify(position.as_dict())

@bp.route('/<int:user_id>/<int:position_id>', methods=['DELETE'])
def remove_position(user_id, position_id):
    position = get_position_by_id(position_id)
    if not position or position.user_id != user_id:
        return jsonify({'error': 'Not found'}), 404
    success = delete_position(position_id)
    if not success:
        return jsonify({'error': 'Not found'}), 404
    return jsonify({'result': 'success'}) 

File: frontend/src/api/userPositionApi.js
import axios from 'axios';
import API_CONFIG from './config';

const api = axios.create({
  baseURL: API_CONFIG.baseURL,
  timeout: API_CONFIG.timeout,
});

export const userPositionApi = {
  // 获取用户持仓列表
  getPositions: (userId, params = {}) => api.get(`/api/user_position/${userId}`, { params }),
  // 获取单条持仓记录
  getPosition: (userId, id) => api.get(`/api/user_position/${userId}/detail/${id}`),
  // 新增持仓记录
  addPosition: (userId, data) => api.post(`/api/user_position/${userId}`, data),
  // 更新持仓记录
  updatePosition: (userId, id, data) => api.put(`/api/user_position/${userId}/${id}`, data),
  // 删除持仓记录
  deletePosition: (userId, id) => api.delete(`/api/user_position/${userId}/${id}`),
}; 